cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -511,3 +511,105 @@ type Arg struct {
+	Patches *Patches
+
+	// PrintConfig specifies whether the configuration file that will be used in the upgrade should be printed or not.
+	PrintConfig *bool
+}
+
+// UpgradeDiffConfiguration contains a list of configurable elements which is specific for "kubeadm upgrade diff" command.
+type UpgradeDiffConfiguration struct {
+	// DiffContextLines is the number of lines of context in the diff.
+	DiffContextLines int
+}
+
+// UpgradeNodeConfiguration contains a list of configurable elements which is specific for "kubeadm upgrade node" command.
+type UpgradeNodeConfiguration struct {
+	// CertificateRenewal instructs kubeadm to execute certificate renewal during upgrades.
+	CertificateRenewal *bool
+
+	// DryRun tells if the dry run mode is enabled, don't apply any change if it is and just output what would be done.
+	DryRun *bool
+
+	// EtcdUpgrade instructs kubeadm to execute etcd upgrade during upgrades.
+	EtcdUpgrade *bool
+
+	// IgnorePreflightErrors provides a slice of pre-flight errors to be ignored during the upgrade process, e.g. 'IsPrivilegedUser,Swap'.
+	// Value 'all' ignores errors from all checks.
+	IgnorePreflightErrors []string
+
+	// SkipPhases is a list of phases to skip during command execution.
+	// The list of phases can be obtained with the "kubeadm upgrade node phase --help" command.
+	SkipPhases []string
+
+	// Patches contains options related to applying patches to components deployed by kubeadm during `kubeadm upgrade`.
+	Patches *Patches
+}
+
+// UpgradePlanConfiguration contains a list of configurable elements which is specific for "kubeadm upgrade plan" command.
+type UpgradePlanConfiguration struct {
+	// AllowExperimentalUpgrades instructs kubeadm to show unstable versions of Kubernetes as an upgrade
+	// alternative and allow upgrading to an alpha/beta/release candidate version of Kubernetes.
+	AllowExperimentalUpgrades *bool
+
+	// Enable AllowRCUpgrades will show release candidate versions of Kubernetes as an upgrade alternative and
+	// allow upgrading to a release candidate version of Kubernetes.
+	AllowRCUpgrades *bool
+
+	// DryRun tells if the dry run mode is enabled, don't apply any change if it is and just output what would be done.
+	DryRun *bool
+
+	// IgnorePreflightErrors provides a slice of pre-flight errors to be ignored during the upgrade process, e.g. 'IsPrivilegedUser,Swap'.
+	// Value 'all' ignores errors from all checks.
+	IgnorePreflightErrors []string
+
+	// PrintConfig specifies whether the configuration file that will be used in the upgrade should be printed or not.
+	PrintConfig *bool
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// UpgradeConfiguration contains a list of elements that is specific "kubeadm upgrade"-only runtime information.
+type UpgradeConfiguration struct {
+	metav1.TypeMeta
+
+	// Apply holds a list of elements that is configurable for "kubeadm upgrade apply" command.
+	Apply UpgradeApplyConfiguration
```

Comment:
these should be pointers.

Reply:
Actually, I have a version to make this a pointer before pushing this update.

I am in favor of non-pointers here, because, 
- pointer here will make the access to the field of the substructure more complicit to handle with, I need make sure this is not a nil pointer and then make sure its field is not a nil pointer as well. e.g. 
If I want to access `UpgradeApplyConfiguration.AllowRCUpgrades`, some check is needed to make sure both `UpgradeApplyConfiguration` and `AllowRCUpgrades` is not nil. The same check is applicable for all those fields.
Some tweak around `ValueFromFlagsOrConfig` has already done [here](https://github.com/kubernetes/kubernetes/pull/114062/files#diff-88bb647764f810575a6c3abf75e2986650cc9b60dbe1de6b7d31fba02106d060R153-R156)


- Almost all the substructure in `types.go` is not defined as pointer, e.g. `InitConfiguration.NodeRegistrationOptions` and `InitConfiguration.APIEndpoint`. I understand it might better for a pointer field if we access it directly to judge whether it's set or not. But I don't think we have too many cases to read it directly, this is somehow different with other components in the Kubernetes where the config in the config file is preferred instead of flags.

Reply:
there is like a tradeoff for both options. one + for pointers is validation. if both substructure A and B are not nil, we can throw an error e.g. both apply and plan. otherwise we have to a compare struct to a default empty struft which is not really possible for some types fileds.

Reply:
also we are talking about separate code paths if we are having the substructures. e.g. all apply code must use the apply configuration and we will know it's not nil.

we can also pass the apply configuration instead of the whole upgrade configuration but i can see that might be more difficult as upgrade is a bit of a mess.

Reply:
> if both substructure A and B are not nil, we can throw an error e.g. both apply and plan.

well, I don't think we should throw an error for this case, it's could be a config with all those config set, each specific command just need to care the config which is related.
One config file to hold all the config for all the commands is the main advantage compared with top-level flatten configuration.  So that you can just provide one config file instead of 4 config file.

Reply:
sgtm, but we need consensus with the others.
i find it irrigating that this was not discussed on a kep or a google doc.

appreciate your efforts, though.